gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/javaversions/BaselineJavaVersion.java

```diff
@@ -101,127 +101,150 @@ public void execute(CheckJavaVersionsTask task) {
     }
 
     private static void configureCompilationTasks(
-            Project project, Property<ChosenJavaVersion> target, Provider<BaselineJavaToolchain> javaToolchain) {
-        project.getTasks().withType(JavaCompile.class).configureEach(new Action<JavaCompile>() {
-            @Override
-            public void execute(JavaCompile javaCompileTask) {
-                javaCompileTask.getJavaCompiler().set(javaToolchain.flatMap(BaselineJavaToolchain::javaCompiler));
-                javaCompileTask
-                        .getOptions()
-                        .getCompilerArgumentProviders()
-                        .add(new EnablePreviewArgumentProvider(target));
-
-                // Set sourceCompatibility to opt out of '-release', allowing opens/exports to be used.
-                javaCompileTask.doFirst(new Action<Task>() {
-                    @Override
-                    public void execute(Task task) {
-                        ((JavaCompile) task)
-                                .setSourceCompatibility(
-                                        target.get().javaLanguageVersion().toString());
-                    }
-                });
-            }
+            Project project,
+            Property<ChosenJavaVersion> target,
+            JavaToolchains baselineConfiguredJavaToolchains,
+            BaselineJavaVersionsExtension rootExtension,
+            JavaToolchainService javaToolchainService) {
+
+        project.getTasks().withType(JavaCompile.class).configureEach(javaCompileTask -> {
+            setJavaCompiler(
+                    javaCompileTask, rootExtension, baselineConfiguredJavaToolchains, javaToolchainService, target);
+            javaCompileTask.getOptions().getCompilerArgumentProviders().add(new EnablePreviewArgumentProvider(target));
+            // Set sourceCompatibility to opt out of '-release', allowing opens/exports to be used.
+            javaCompileTask.doFirst(task -> ((JavaCompile) task)
```

Comment:
TaskActions must be written using `new Action` rather than lambas, otherwise Gradle will make this task never up-to-date. It can't get a serialised name for a lambda, but can for a task. There is an errorprone for this, but errorprone does not run in here as for some reason it's using a groovy source.

Reply:
so then we need to use Action here as well: https://github.com/palantir/gradle-baseline/pull/2783/files#diff-bd31167bf1fd132bca9a139faa8fc09fb838bc5067c668f99b116be494a0e8b6R104 right ? as this is a taskAction

Reply:
Yep, anything given to doFirst/doLast.